### PR TITLE
fix: SetLogger via klog.SetLogger will output an unexpected newline

### LINKED ISSUE #370: An extra line break was output
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

This is a log without SetLogger 

```
I0217 18:17:05.411664   62273 shared_informer.go:280] Caches are synced for project-controller
I0217 18:17:05.411821   62273 shared_informer.go:280] Caches are synced for application-controller
```

This is set by SetLogger 

```
2023-02-17T18:18:12.680+0800    INFO    client-go       Waiting for caches to sync for application-controller

2023-02-17T18:18:12.680+0800    INFO    client-go       Waiting for caches to sync for project-controller

2023-02-17T18:18:12.781+0800    INFO    client-go       Caches are synced for project-controller

2023-02-17T18:18:12.781+0800    INFO    client-go       Caches are synced for application-controller


```

They output a new line between them. After my investigation, I found that klog adds a new line at the end of the msg, but for the log output by the logr, the new line will be automatically added.

He can work normally under stdout, because stdout needs this newline

https://github.com/kubernetes/klog/blob/main/klog.go#L729

**What did you expect to happen:**

 Do not add new lines when using logr 

```
2023-02-17T18:18:12.680+0800    INFO    client-go       Waiting for caches to sync for application-controller
2023-02-17T18:18:12.680+0800    INFO    client-go       Waiting for caches to sync for project-controller
2023-02-17T18:18:12.781+0800    INFO    client-go       Caches are synced for project-controller
2023-02-17T18:18:12.781+0800    INFO    client-go       Caches are synced for application-controller
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by aimuz**
klog always adds a newline to the msg. klog will work fine without klog.

Set logr with klog.SetLogger, and klog will also pass the msg with the newline added to logr, which will result in an accidental newline being added.

step1: klog.Info("hello world")
step2: msg = msg + "\n"
step3: stdout.W(msg) or logr.Info(msg[:len(msg)-1])

fixes #370

Signed-off-by: aimuz <mr.imuz@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by pohly** [comment]
@aimuz: to reproduce the test failures, use `go test ./... && (cd examples && go test ./...)`.

**Comment by aimuz** [comment]
thanks you @pohly

**Comment by aimuz** [comment]
PR is ready, please feel free to review

File: klog.go [hunk 1]
```diff
@@ -873,6 +873,9 @@ func (l *loggingT) output(s severity.Severity, logger *logWriter, buf *buffer.Bu
 		if logger.writeKlogBuffer != nil {
 			logger.writeKlogBuffer(data)
 		} else {
+			if data[len(data)-1] == '\n' {
```
Comment (pohly) [inline klog.go:876]: I know, `data` can probably never be empty, but perhaps check for it anyway?

Comment (aimuz) [reply klog.go]: Yes, it is necessary that

done

**Review by pohly — CHANGES_REQUESTED** [review]
Looks good. Can you squash into one commit with a useful commit message?

**Comment by pohly** [comment]
The first commit message is already okay, no need to edit it.

**Comment by aimuz** [comment]
done

**Review by pohly — APPROVED** [review]
/lgtm
/approve

**Comment by aimuz** [comment]
/assign @dims 

For approval.

**Pull request merged**